### PR TITLE
Only compute new_untracked_symbols and new_unbacked_bindings if needed.

### PR DESCRIPTION
**Pull request by laithsakka**
Summary:
237s -> 198.. 
buck2 run fbcode//mode/opt fbcode//torchrec/distributed/tests:pt2_compile_benchmark -- --num-features=2000

Test Plan: NA

Differential Revision: D65638637

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/fx/passes/runtime_assert.py [hunk 1]
```diff
@@ -350,27 +350,29 @@ def match_symbol(symint, cb):
                 and (sym_expr := _get_sym_val(node)) is not None
             ):
                 # this guards against deleting calls like item() that produce new untracked symbols
-                new_untracked_symbols = sym_expr.free_symbols - expr_to_proxy.keys()
+                def new_untracked_symbols():
+                    return sym_expr.free_symbols - expr_to_proxy.keys()
```
Comment (isuruf) [inline torch/fx/passes/runtime_assert.py:354]: Maybe do
```suggestion
                def has_new_untracked_symbols():
                    return bool(sym_expr.free_symbols - expr_to_proxy.keys())
```
to make it clear what it's used for

**Review by isuruf — APPROVED** [review]

**Comment by laithsakka** [comment]
mad it even better.

File: torch/fx/passes/runtime_assert.py [hunk 2]
```diff
@@ -350,27 +350,36 @@ def match_symbol(symint, cb):
                 and (sym_expr := _get_sym_val(node)) is not None
             ):
                 # this guards against deleting calls like item() that produce new untracked symbols
-                new_untracked_symbols = sym_expr.free_symbols - expr_to_proxy.keys()
+                def has_new_untracked_symbols():
+                    for symbol in sym_expr.free_symbols :
+                        if symbol not in expr_to_proxy.keys():
```
Comment (isuruf) [inline torch/fx/passes/runtime_assert.py:355]:
```suggestion
                        if symbol not in expr_to_proxy:
```

**Review by anijain2305 — APPROVED** [review]

**Pull request closed without merging**